### PR TITLE
[FLINK-24161] Fix interplay of stop-with-savepoint w/o drain with final checkpoints

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change
* In endData() if there is an pending savepoint without drain, we fail the task by throw an exception.
* When triggering stop-with-savepoint without drain, if endOfDataReceived = true , we then return false to reject the savepoint.

Both of the above 2 cases cause a failover. 

## Verifying this change
Added tests in StreamTaskFinalCheckpointsTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 1]
```diff
@@ -529,6 +530,13 @@ protected void processInput(MailboxDefaultAction.Controller controller) throws E
     }
 
     protected void endData() throws Exception {
+
+        if (syncSavepointWithoutDrain != null) {
```
Comment (gaoyunhaii) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:534]: Perhaps we only do the check if checkpoints after tasks finished are enabled~? If not, the task would not wait for the final checkpoint and would not be blocked, thus we might not change the behavior if flip-147 is not enabled~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 2]
```diff
@@ -552,6 +560,7 @@ protected void setSynchronousSavepoint(long checkpointId, boolean isDrain) {
                 syncSavepointWithDrain = checkpointId;
             }
         } else {
+            if (endOfDataReceived) {}
```
Comment (gaoyunhaii) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:563]: This line should be able to be removed~

**Review by gaoyunhaii — COMMENTED** [review]
Hi @dawidwys very thanks for opening the PR! I left some comments~ I'll also have a test with the original issue with the pr patched~

File: .idea/vcs.xml [hunk 3]
```diff
@@ -27,7 +11,7 @@
           <option name="issueRegexp" value="FLIP\-\d+" />
           <option name="linkRegexp" value="https://cwiki.apache.org/confluence/display/FLINK/$0" />
         </IssueNavigationLink>
-          <IssueNavigationLink>
+        <IssueNavigationLink>
```
Comment (gaoyunhaii) [inline .idea/vcs.xml:14]: This seems to be included by accident?

**Comment by gaoyunhaii** [comment]
I do the manual check, the savepoint would fail as expected and cause failover, then the job could be finished with stop-with-savepoint normally.

File: .idea/vcs.xml (hunk 3 above)
Comment (dawidwys) [reply .idea/vcs.xml:14]: argh... yes, sorry for that

**Review by gaoyunhaii — APPROVED** [review]
Very thanks @dawidwys for opening the PR! LGTM and +1 to merge~

**Pull request closed without merging**